{% extends 'base.html' %}
{% import 'macros.html' as macros %}

<!-- Title -->
{% block title %} BioJupies | Generate Signature {% endblock %}

<!-- Content -->
{% block content %}

<div class="px-6">

	<!-- Title -->
	<div class="row pt-4">
		<div class="col-12 very-large text-center light px-5">What samples would you like to compare?</div>
	</div>

	<!-- Introduction -->
	<hr width="100%" class="my-4">
	<div class="row">
		<div class="light very-small my-1 col-lg-8 col-xl-9 text-justify">
			One or more of the selected tools require generating a <span class="bold">gene expression signature<sup>
				<i class="fa fa-question-circle" data-template='<div class="popover signature-popover micro text-justify" role="tooltip"><div class="arrow"></div><h3 class="popover-header"></h3><div class="popover-body"></div></div>' data-toggle="popover" data-html="true" data-trigger="hover" title="What is a gene expression signature?" data-content="<div>Gene expression signatures are <b>alterations in the patterns of gene expression</b> that occur as a result of cellular perturbations such as drug treatments, gene knock-downs or diseases.</div><div class='mt-2'>They can be generated by <b>comparing gene expression across two groups of samples</b>, typically a control group and a treated group.</div><div class='mt-2'>Generated signatures consist of a <b>ranked list of genes</b>, ordered by a metric which indicates the direction and magnitude of their differential expression (typically, a fold-change or a p-value).</div>"></i></sup></span>.
				To generate one, you must <b>define two groups of samples</b> whose gene expression you wish to compare by using the form below.
				Once you have defined the desired groups, click <b>Continue</b> to proceed.
			<!-- Use the form below to <span class="bold">add or remove data analysis and visualization tools</span> to your notebook. -->
			<!-- These tools will analyze the selected dataset and embed interactive results in your notebook. -->
			<!-- Once you have selected the desired tools, click <b>Continue</b> to proceed. -->
		</div>
		<div class="col-lg-4 col-xl-3 my-auto text-center pt-3 pt-lg-0">
			<button class="btn black border-custom bg-white nodecoration mr-2" onclick="goBack();"><i class="fa fa-angle-left mr-2"></i>Back</button>
			<button class="btn navigate white border-custom bg-blue nodecoration" form="configure-signature-form" type="submit" formaction="{{ url_for('configure_analysis') }}" formmethod="post">
				Continue<i class="fa fa-angle-right ml-2"></i>
			</button>
		</div>
	</div>

	<!-- Configure Signature Form -->
	<div class="row pb-5 pt-2">
		<div class="col-12 px-3">
			<form id="configure-signature-form">


				<!-- Define Groups -->
				<div class="row px-3">

					<!-- Name Groups -->
					<div class="medium light mt-2 mb-2">What are the names of the groups?</div>
					<div class="very-small light mb-3">
						First, <b>name the groups of samples</b> you wish to compare using the text boxes below.
						It is recommended to use names descriptive of the groups' experimental condition: for example, <span class="font-italic">WT, DMSO, Vehicle</span> for the <b>Control</b> group; <span class="font-italic">Metastatic, p53 KO, Dasatinib-6h</span> for the <b>Perturbation</b> group.
					</div>

					<!-- Group Labels -->
					<div class="col-lg-9 col-xl-8 pl-0 mt-1">
						<input type="text" name="group_a_label" class="group-label-input rounded-left border-custom border-right-0 px-2 py-1 light very-small nodecoration" data-group="a" value="Control" maxlength="50">
						<span class="py-1 border-custom rounded-right d-inline-block ml-neg px-2 font-italic" data-toggle="tooltip" title="The number of samples assigned to the group. To improve the statistical validity of the gene expression signature, the group must have at least three samples." data-group="a"><span class="sample-counter">0</span> samples</span>
						<span class="ml-2 mr-1 d-inline-block">vs</span>
						<div class="d-block d-lg-inline-block mt-2 mt-lg-0">
							<input type="text" name="group_b_label" class="group-label-input rounded-left border-custom border-right-0 px-2 py-1 light very-small nodecoration" data-group="b" value="Perturbation" maxlength="50">
							<span class="sample-counter py-1 border-custom rounded-right d-inline-block ml-neg px-2 font-italic" data-toggle="tooltip" title="The number of samples assigned to the group. To improve the statistical validity of the gene expression signature, the group must have at least three samples." data-group="b"><span class="sample-counter">0</span> samples</span>
						</div>
						<input type="hidden" value="Select..." class="group-label-input" data-group="none" maxlength="50">
					</div>

					<!-- Group Prediction -->
					<div class="col-lg-3 col-xl-4 my-auto pt-3 pt-lg-0">
						<div class="d-table">
							<div class="d-table-cell">
								<div class="onoffswitch">
								    <input type="checkbox" name="onoffswitch" class="onoffswitch-checkbox" id="predict-groups" checked>
								    <label class="onoffswitch-label" for="predict-groups"></label>
								</div>
							</div>
							<div class="d-table-cell pl-2 very-small regular">
								Predict Groups <sup><i class="fa fa-question-circle" data-toggle="tooltip" data-placement="right" title="Automatically predicts the samples belonging to each group based on the assigned names."></i></sup>
							</div>
						</div>
					</div>
				</div>

				<!-- Assign Samples -->
				<div class="row px-3 mt-2">

					<!-- Sample Selection -->
					<div class="medium light mt-4 mb-2">Which samples belong to the groups?</div>
					<div class="very-small light mb-4 text-justify">
						Second, <b>select the samples</b> you wish to assign to the groups by using the dropdown menus below.
						To improve the statistical validity of the gene expression signature, each group must have <b>at least three samples</b>.
					</div>

					<!-- Sample Table -->
					<table id="sample-table" class="stripe">
						<thead>
							<tr>
								<th class="text-center very-small regular">Group</th>
								{% for col in j.columns %}
									<th class="text-center very-small regular">{{ col.title() }}</th>
								{% endfor %}
							</tr>
						</thead>
						<tbody>
							{% for index, rowData in j.iterrows() %}
								<tr class="sample-row very-small light text-center px-2 py-1" data-group="none">
									<td>
										<div class="dropdown group-dropdown">
											<button class="btn btn-secondary dropdown-toggle black border-black nodecoration regular very-small pr-2" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false" data-group="none">Select...</button>
											<div class="dropdown-menu">
												<span class="dropdown-item" data-dropdown-group="a"><div class="dropdown-square border-grey rounded cursor-pointer"></div><span class="bg-transparent cursor-pointer" data-group="a">Control</span></span>
												<span class="dropdown-item" data-dropdown-group="b"><div class="dropdown-square border-grey rounded cursor-pointer"></div><span class="bg-transparent cursor-pointer" data-group="b">Perturbation</span></span>
												<span class="dropdown-item" data-dropdown-group="none"><div class="dropdown-square border-grey rounded cursor-pointer"></div><span class="bg-transparent cursor-pointer" data-group="none">None</span></span>
											</div>
											<input type="hidden" name="{{rowData.get('accession') if rowData.get('accession') else rowData.get('sample') }}-group" value="none">
										</div>
									</td>
									{% for col in j.columns %}
										<td>{{ rowData[col] }}</td>
									{% endfor %}
								</tr>
							{% endfor %}
						</tbody>
					</table>
				</div>

				<!-- Previous Form Data -->
				{% for key, values in f.lists() %}
					{% for value in values %}
						<input type="hidden" name="{{ key }}" value="{{ value }}">
					{% endfor %}
				{% endfor %}
				<input type="hidden" name="requires_signature" value="False">

			</form>
		</div>
	</div>
</div>

{% endblock %}

{% block footer %}

<script type="text/javascript">

	$(document).ready(function(){

		// Data Table
	    $('#sample-table').DataTable({
	    	"lengthMenu": [[-1, 10, 25, 50], ["All", 10, 25, 50]],
	    	  "columnDefs": [
				{ "orderable": false, "targets": 0 },
				{ "searchable": false, "targets": 0 },
			  ]
	    });

	    // Manual Class Selection
	    $('.dropdown-item').click(function(evt) {

	    	// Update Dropdown
	    	var $item = $(evt.target).hasClass('dropdown-item') ? $(evt.target) : $(evt.target).parents('.dropdown-item'),
	    		new_group = $item.data('dropdown-group'),
	    		new_label = $('.group-label-input[data-group="'+new_group+'"]').val();
    		$(evt.target).parents('.dropdown').find('button').html(new_label+'&nbsp').attr('data-group', new_group);
    		$(evt.target).parents('.dropdown').find('input').val(new_group);
    		$(evt.target).parents('tr').attr('data-group', new_group);

    		// Update Counter
    		$('[data-group="a"] .sample-counter').html($('.group-dropdown button[data-group="a"]').length);
    		$('[data-group="b"] .sample-counter').html($('.group-dropdown button[data-group="b"]').length);
	    })

	    // Automated Class Selection
	    function autoSelect(group, value) {
	    	var hit = false;
	    	if (value.length > 0) {
		    	$('#sample-table tr').each(function(index, row) {
		    		if ($(row).find('td:not(:first-child)').text().toLowerCase().indexOf(value.toLowerCase()) > -1) {
		    			$(row).find('[data-dropdown-group="'+group+'"]').click();
		    			hit = true;
		    		} else {
		    			if (group == $(row).find('.dropdown-toggle').attr('data-group')) {
			    			$(row).find('[data-dropdown-group="none"]').click();
		    			}
		    		}
		    	})
	    	}
	    	return hit
	    }

	    $('.group-label-input').change(function(evt) {
	    	var value = $(evt.target).val(),
	    		group = $(evt.target).data('group');
	    	$('[data-group="'+group+'"]:not(tr):not(span)').text(value);
	    	if ($('#predict-groups').prop('checked')) {
		    	autoSelect(group, value);
	    	}
	    })

	    // Predict Control Classes
	    var hit, terms=['Control', 'Ctl', 'Untreated', 'Vehicle', 'Normal', 'Non-specific', 'DMSO', 'WT'];
	    for (i = 0; i < terms.length; i++) {
	    	hit = autoSelect('a', terms[i]);
	    	if (hit) { 
	    		$('input[name="group_a_label"]').val(terms[i]);
	    		$('[data-group="a"]:not(tr)').text(terms[i]);
	    		break
	    	}
	    }

	    // Remove Predicted Groups
	    $('#predict-groups').change(function(evt) {
	    	if (!$('#predict-groups').prop('checked')) {
	    		$('[data-dropdown-group="none"]').click();
	    	} else {
	    		autoSelect('a', $('[name="group_a_label"]').val());
	    		autoSelect('b', $('[name="group_b_label"]').val());
	    	}
	    })

	    // Check Group Size on Submission
	    $('#configure-signature-form').submit(function(evt) {
	    	var a_samples = $('#sample-table button[data-group="a"]').length,
	    		b_samples = $('#sample-table button[data-group="b"]').length;
	    	if (a_samples < 3) {
		    	evt.preventDefault();
		    	alert('Please add more samples to the '+$('[name="group_a_label"]').val()+' group.\n\nGroups are required to contain at least 3 samples. However, the group has '+a_samples+'.')
	    	} else if (b_samples < 3) {
		    	evt.preventDefault();
		    	alert('Please add more samples to the '+$('[name="group_b_label"]').val()+' group.\n\nGroups are required to contain at least 3 samples. However, the group has '+b_samples+'.')

	    	}
	    })

	});

</script>

{% endblock %}
