{% extends 'base.html' %}
{% import 'macros.html' as macros %}

<!-- Title -->
{% block title %} BioJupies | Upload Matrix File {% endblock %}

<!-- Head -->
{% block head %}
<style type="text/css">
	.tooltip-inner {
		max-width: 600px;
	}
</style>
{% endblock %}

<!-- Content -->
{% block content %}

<div class="px-6 mb-5">

	<!-- Title -->
	<div class="row pt-4">
		<div class="col-12 very-large text-center light px-5">What groups do the samples belong to?</div>
	</div>

	<!-- Introduction -->
	<hr width="100%" class="my-4">
	<div class="row">
		<div class="light very-small my-1 col-lg-8 col-xl-9 text-justify">
			The uploaded file contains <b>{{sample_groups|length}} samples.</b>
			To improve the quality of the analyses, it is required to <b>define groups of samples</b> based on their experimental conditions.
			To do this, manually enter the groups or upload a sample metadata file using the form below.
			Once you have defined the desired groups, click <b>Continue</b> to proceed.
		</div>
		<div class="col-lg-4 col-xl-3 my-auto text-center pt-3 pt-lg-0">
			<button class="btn black border-custom bg-white nodecoration mr-2" onclick="goBack();"><i class="fa fa-angle-left mr-2"></i>Back</button>
			<button id="continue-button" class="btn navigate white border-custom bg-blue nodecoration">Continue<i class="fa fa-angle-right ml-2"></i></button>
		</div>
	</div>

	<!-- Add Metadata -->
	<div class="row">
		<hr width="100%" class="mt-3 mb-0">
		<div class="col-12 px-4">

			<!-- Add introduction -->
			<div id="info" class="text-justify collapse px-1">
				<div class="mt-4 light">Incorporating <b>prior information regarding samples</b>, such as the experimental condition or tissue of origin, allows to extract more knowledge from gene expression data.</div>
				<div class="mt-3 light">Using the form below, you can add information about the samples in the uploaded dataset, such as the experimental condition or tissue of origin.</div>
				<div class="mt-3 light">To achieve this, fill in the form on the left, or upload a metadata file using the form on the right. Once this process is complete, click continue to proceed. </div>
				<div class="pt-4 pb-2">
					<hr width="100%" class="my-0">
				</div>
			</div>

		</div>
	</div>
	<div id="options" class="row mt-2 px-2">

		<!-- Manual Annotation Table -->
		<div class="col-6 px-0">

			<!-- Predict Groups -->
			<div class="hidden">
				<div class="d-table-cell">
					<div class="onoffswitch">
					    <input type="checkbox" name="onoffswitch" class="onoffswitch-checkbox" id="predict-groups" checked>
					    <label class="onoffswitch-label" for="predict-groups"></label>
					</div>
				</div>
				<div class="d-table-cell pl-2 very-small regular">
					Predict Groups <sup><i class="fa fa-question-circle" data-toggle="tooltip" data-html="true" data-placement="right" title="Automatically predicts the<br>groups based on the<br>sample names."></i></sup>
				</div>
			</div>

			<!-- Title and Description -->
			<div class="light medium mt-2 mb-2">Manually Add the Groups</div>
			<div class="light very-small mb-3"><b>Manually specify the sample groups</b> using the text boxes below.</div>

			<!-- Sample Annotation Table -->
			<table id="sample-annotation-table" class="tiny text-center table-striped border-grey w-100 mb-5">
				<thead class="very-small">
					<tr class="border-grey border-left-0 border-right-0">
						<th>Sample<sup><i class="fa fa-question-circle small ml-1 micro" data-toggle="tooltip" data-html="true" data-placement="right" title="The names of the samples<br>in the uploaded gene<br>expression file."></i></sup></th>
						<th class="px-3 py-1">Group<sup><i class="fa fa-question-circle small ml-1 micro" data-toggle="tooltip" data-html="true" data-placement="right" title="
									<div class='text-left px-2 py-1 light'>
										The group to which each sample belongs to, based on its experimental condition or biological source.
										<div class='mt-3 px-2'>
											Examples:
											<ul class='pl-4'>
												<li>For <b>genetic perturbations</b>: <span class='font-italic'>WT, KO, KD, Overexpression</span>.</li>
												<li>For <b>small molecule perturbations</b>: <span class='font-italic'>Vehicle, Cytarabine 5h, Valdecoxib 50µM</span>.</li>
												<li>For <b>diseases</b>: <span class='font-italic'>healthy control, primary tumor, nephropathy</span>.</li>
											</ul>
										</div>
									</div>">
								</i>
							</sup>
						</th>
					</tr>
				</thead>
				<tbody>
					{% for sample in sample_groups %}
						<tr>
							<td class="px-3 py-2 bold tiny" style="max-width: 350px; overflow-x: auto;">{{ sample['sample'] }}</td>
							<td class="px-3 py-2"><input class="form-control nodecoration tiny" type="text" name="{{sample}}-group"></td>
							<!-- <td class="px-3 py-2"><input class="form-control nodecoration tiny" type="text" name="{{sample}}-group" value="{{ sample['group'] }}" data-predicted-group="{{ sample['group'] }}"></td> -->
						</tr>
					{% endfor %}
				</tbody>
			</table>
		</div>

		<!-- OR Separator -->
		<div class="col-1 text-center px-0 pt-5 mt-5">
			<span>OR</span>
		</div>

		<!-- Metadata File Upload -->
		<div class="col-5 px-0">

			<!-- Title and Description -->
			<div class="light medium mt-2 mb-2">Upload a Metadata File</div>
			<div class="light very-small mb-3"><b>Upload a sample metadata file</b><sup>
				<i class="fa fa-question-circle" data-template='<div class="popover info-popover micro text-justify" role="tooltip"><div class="arrow"></div><h3 class="popover-header"></h3><div class="popover-body"></div></div>' data-toggle="popover" data-html="true" data-trigger="hover" title="What is a sample metadata file?" data-content="<div>Sample metadata files <b>contain information describing the samples in the dataset</b>, such as the experimental conditions they have been treated with or their biological sources.</div><div class='mt-2'>.</div><div class='mt-2'>.</div>"></i></sup></span> using the form below.</div>

			<!-- Dropzone -->
			<div id="dropzone" class="dropzone border-grey rounded mt-3" style="background-image: url({{ url_for('static', filename='img/dropzone/metadata.png') }}); background-size: 100% auto;"></div>
			<div class="tiny mt-2">
				Supported formats: <span class="font-italic">.txt, .csv, .tsv, .xls, .xlsx</span>.
				<div class="dropdown example-dropdown show float-right">
					<a class="btn bg-white border-custom black nodecoration tiny regular my-auto dropdown-toggle" href="#" role="button" id="dropdownMenuLink" data-toggle="dropdown">
						Example
					</a>
					<div class="dropdown-menu" aria-labelledby="dropdownMenuLink">
						<a class="dropdown-item" href="#">Load Example</a>
						<a class="dropdown-item" href="{{ url_for('static', filename='data/biojupies_example_metadata.txt') }}" download>Download Example File</a>
					</div>
				</div>
			</div>
		</div>
	</div>


	<!-- Uploaded File Preview -->
	<div class="row">
		<div id="preview" class="col-12 d-none mt-1 mb-5">
			<div class="light medium mt-2 mb-2">Sample Groups</div>
			<div class="light very-small mb-3">The table below displays a preview of the uploaded sample groups file:</div>
		</div>
	</div>

	<form id="upload-metadata-form" action="{{ url_for('upload_'+uploadtype) }}" method="post">
		<input type="hidden" id="expression" name="expression" value="{{f.get('expression')}}">
		<input type="hidden" id="metadata" name="metadata" value="">
		{% if f.get('alignment_uid') %}
			<input type="hidden" name="alignment_uid" value="{{ f['alignment_uid'] }}">
		{% endif %}
		{% if f.get('reference_genome') %}
			<input type="hidden" name="reference_genome" value="{{ f['reference_genome'] }}">
		{% endif %}
		{% if f.get('qc') %}
			<input type="hidden" name="qc" value="{{ f['qc'] }}">
		{% endif %}

		<!-- Name Dataset -->
		{% if current_user.is_authenticated %}
			<div class="row">
				<div class="col-12">
					<div class="light medium mb-2">Name your Dataset</div>
					<div class="light very-small mb-3"><b>Optionally, insert a name</b> of your dataset. The name will be displayed in your <a href="{{ url_for('dashboard') }}" target="_blank">user dashboard</a> and can be changed at any time.</div>
					<input type="text" value="Untitled Dataset" class="form-control" name="dataset_title">
				</div>
			</div>
		{% endif %}

	</form>
		
</div>

{% endblock %}


{% block footer %}

<script type="text/javascript">

// Add Table
var addTable = function(dataset) {

	// Interfaces
	$('#options').addClass('hidden');
	$('#sample-annotation-table').addClass('hidden');
	$('#intro').html('The uploaded metadata file contains <span class="highlight">'+dataset['columns'].length+' fields</span> and <span class="highlight">'+dataset['index'].length+' rows</span>. Check that the preview below is correct, then click Continue to proceed.');

	// Add Table
	addPreviewTable(dataset, false);
}

// Dropzone handler
$("#dropzone").dropzone({ 
	url: "{{ url_for('upload_dataframe_api') }}",
	acceptedFiles: ".txt,.csv,.tsv,.xls,.xlsx",
	success: function(file, response){

		// Parse response
		var response = JSON.parse(response);

		// Check if sample names match
		var expression_sample_names = JSON.parse($('#expression').val())['columns'],
			metadata_sample_names = response['index'],
			extra_names = metadata_sample_names.diff(expression_sample_names),
			missing_names = expression_sample_names.diff(metadata_sample_names);

		// Remove extra names
		if (extra_names.length) {
			$.each(extra_names, function(i, extra_name) {
				var extra_index = metadata_sample_names.indexOf(extra_name);
				response['index'].splice(extra_index, 1);
				response['data'].splice(extra_index, 1);
			})
		}

		// Alert if missing names
		if (missing_names.length) {
			alert('The following samples are missing from the metadata file:\n\t• '+missing_names.join('\n\t• ')+'\n\nPlease make sure that the rows of the metadata file match the number samples in the expression file.');
			console.log('missing');
		} else {

			// Add table
			addTable(response);
		}
	}
});


// Load Example
$('.example-dropdown .dropdown-item:first-child').click(function(evt) {

	// AJAX Query
	$.ajax({
		url: "{{ url_for('example_table_api') }}",
		method: 'post',
		data: JSON.stringify({'filename': 'biojupies_example_metadata.txt'}),
		contentType: "application/json; charset=utf-8",
		dataType: 'json',
		success: function(response) {

	    	// Add dataset
	    	addTable(response);

		}
	})
})

// Check if all rows of input have been completed
$('#sample-annotation-table input').change(function(evt) {
	var complete = true;
	$('#sample-annotation-table input').each(function(i, input) {
		if ($(input).val().length === 0) {
			complete = false;
		}
	})
})

// Submit metadata
$('#continue-button').click(function(evt) {

	// Get available table
	var $table = $('table:not(.hidden)'),
		metadata = serializeTable($table);

	// Remove tooltip
	metadata[0][0] = metadata[0][0].split('<sup')[0];
	metadata[0][1] = metadata[0][1].split('<sup')[0];

	// Convert to object
	var missing_samples = [];
	$.each(metadata, function(index, elem){
		if ((index > 0) & (elem[1].length === 0)) {
			missing_samples.push(elem[0]);
		}
	})

	// Check groups
	if (missing_samples.length) {
		alert('Please assign a group to the following samples:\n - '+missing_samples.join('\n - '))
	} else {
		$('#metadata').val(JSON.stringify(metadata));
		$('#upload-metadata-form').submit();
	}
})

</script>

{% endblock %}
